jenkins/Pipeline.gy

```diff
@@ -46,34 +49,18 @@ pipeline {
         }
         stage('Build the Pbench Server Container') {
             steps {
-                sh 'buildah login -u="${PODMAN_INTERNAL_USR}" -p="${PODMAN_INTERNAL_PSW}" ${PB_INTERNAL_CONTAINER_REG}'
-                sh 'RPM_PATH=${WORKSPACE_TMP}/rpmbuild/RPMS/noarch/pbench-server-*.rpm bash -ex ./server/pbenchinacan/container-build.sh'
-                sh 'buildah push localhost/pbench-server:${PB_SERVER_IMAGE_TAG} ${PB_SERVER_IMAGE}:${PB_SERVER_IMAGE_TAG}'
+                sh 'buildah login -u="${PODMAN_INTERNAL_USR}" -p="${PODMAN_INTERNAL_PSW}" ${PB_CONTAINER_REG}'
+                sh 'RPM_PATH=${WORKSPACE_TMP}/rpmbuild/RPMS/noarch/pbench-server-*.rpm bash -ex ./server/pbenchinacan/container-build-ci.sh'
+                sh 'buildah push localhost/pbench-server-ci:${PB_SERVER_IMAGE_TAG} ${PB_CONTAINER_REG}/${PB_REPO_NAME}/${PB_SERVER_IMAGE_NAME}:${PB_SERVER_IMAGE_TAG}'
             }
         }
         stage('Deploy server and run functional tests') {
             steps {
-                sh 'envsubst < server/pbenchinacan/pod-nodata.yml | podman play kube -'
-                sleep 30
-                // Use curl to check when server is responding to API requests
-                retry(count: 10) {
-                    sleep 10
-                    sh 'curl --silent localhost:8080/api/v1/endpoints >/dev/null'
-                }
-                sh 'EXTRA_PODMAN_SWITCHES="--network host" jenkins/run server/exec_functests http://localhost:8080'
-                sh 'podman pod stop ${PB_POD_NAME}'
+                sh 'jenkins/run-server-func-tests --cleanup'
```

Comment:
This bothers me a bit, but it's hard to articulate exactly why.

1. It's a bit odd to have this "server" script under `jenkins` rather than under `server`.
2. If we are ever going to run this manually, I can see the advantage of being able to disable the cleanup so we can inspect the droppings: but I'd prefer to have your cleanup `trap` be the default since it's almost always what we'd want.
3. We're going to have agent functional tests, and ideally the server and agent functional test suites will overlap so that we can test agent code talking to our server... which means either we're standing everything up twice, or we have to overlap the two. That's certainly not impossible with two separate self-contained scripts, but it gets a bit messier to follow, while having the combined inline standup and teardown makes the sequence relatively obvious.

Reply:
> * It's a bit odd to have this "server" script under `jenkins` rather than under `server`.

True.

> * If we are ever going to run this manually, I can see the advantage of being able to disable the cleanup so we can inspect the droppings: but I'd prefer to have your cleanup `trap` be the default since it's almost always what we'd want.

I concur -- I think it would be better to have it clean up by default, with an option to skip it.

> * We're going to have agent functional tests, and ideally the server and agent functional test suites will overlap so that we can test agent code talking to our server... which means either we're standing everything up twice, or we have to overlap the two. That's certainly not impossible with two separate self-contained scripts, but it gets a bit messier to follow, while having the combined inline standup and teardown makes the sequence relatively obvious.

AFAIK, the Agent only talks to the Server in a `PUT` operation, so there's not all that much overlap.  (But, yes, it's a _very_ important overlap...for which the Agent test will need a server...and, yes, it seems silly to stand one up twice.)

It seems to me that there are three rather distinct steps to be done:  stand up and stablize the Pbench Server; run tests (presumably Agent **and** Server); and, shut everything down.  It makes sense to do the steps in the Groovy file, except that it makes it hard for a developer to run them; spinning it all off into a script has some advantages, except that it "crosses the streams" between the Agent and the Server (i.e., how do we package it and which directory does it live in?).

So, the framework that we extract from the Groovy file should probably live in a script in the `jenkins` directory; in turn, though, it should invoke scripts which come from the Agent and Server subtrees for the actual testing.

Reply:
See PR #3112.